### PR TITLE
Add bokeh plots and deprecate contributions_pie

### PR DESCRIPTION
**Pull request by raoulcollenteur**
# Short Description

This PR adds basic plotting using Bokeh to Pastas models. To use this, bokeh plots needs to be registered as an extension, as implemented by @dbrakenhoff for plotly. I started with two basic plots, but this can grow over time.

I also added the deprecation for contributions_pie. 

# Checklist before PR can be merged:
- [X] closes issue #638, #711
- [x] is documented
- [x] Format code with [Black formatting](https://black.readthedocs.io)
- [x] type hints for functions and methods
- [x] tests added / passed
- [x] Example Notebook (for new features)
- [x] Remove output for all notebooks with changes

### CODE REVIEW

**Comment by raoulcollenteur** [comment]
@dbrakenhoff shall I extend the plotly notebook example and add bokeh there as well, and rename to interactive_plots.ipynb?

**Comment by raoulcollenteur** [comment]
@dbrakenhoff this PR is ready for Review. But perhaps you can have a quick look at the Plotly plots in the notebook, somehow they don't show up on my system? 

Cheers,
Raoul

**Comment by dbrakenhoff** [comment]
@raoulcollenteur, both are working for me in jupyter lab. Which versions are you on? For me. the following worked:
- plotly 5.16.1
- bokeh 3.4.0
- jupyter lab 4.1.1

FYI, there seems to be some funkiness going on in the bokeh plots:
image

And a note for anyone trying to view the bokeh plots in a VS Code notebook environment, the interactive figures don't show up.
https://github.com/bokeh/bokeh/issues/10765

File: pastas/plotting/bokeh.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+            source=source,
+            legend_label="Observations",
+            color="black",
+            alpha=0.7,
+        )
+        p.line(
+            "index",
+            "Simulation",
+            source=source,
+            legend_label=r"Simulation (R$$^2$$ = {:.2f})".format(rsq),
```
Comment (dbrakenhoff) [inline pastas/plotting/bokeh.py:106]: The superscript isn't formatting nicely for me.

Comment (raoulcollenteur) [reply pastas/plotting/bokeh.py:106]: Same here, I'll have a look.

File: pastas/plotting/bokeh.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            width=int(0.75 * width),
+            y_range=p.y_range,
+            x_axis_type="datetime",
+            y_axis_type=None,
+            tools="",
+            toolbar_location=None,
+            background_fill_color="#efefef",
+        )
+
+        range_tool = RangeTool(x_range=p.x_range)
```
Comment (dbrakenhoff) [inline pastas/plotting/bokeh.py:289]: Very cool :)

**Review by dbrakenhoff — COMMENTED** [review]
See earlier comment about the funkiness in Bokeh plot. Otherwise looking good! 

Setting up somewhat complicated shared axes seems a lot easier in bokeh than in plotly (not having to dive down into the low-level API). And I'm a fan of the range slider at the bottom :). Maybe we can just leave the background of that plot white, so the selected range (also gray) is clearer.

File: pastas/plotting/bokeh.py (hunk 1 above)
Comment (raoulcollenteur) [reply pastas/plotting/bokeh.py:106]: https://docs.bokeh.org/en/3.4.0/docs/user_guide/styling/mathtext.html Not yet supported for legend_labels. So I removed it.

File: pastas/plotting/bokeh.py (hunk 2 above)
Comment (dbrakenhoff) [reply pastas/plotting/bokeh.py:289]: Hm, should add that to the plotly plot as well maybe :innocent: 
https://plotly.com/python/range-slider/

Comment (raoulcollenteur) [reply pastas/plotting/bokeh.py:291]: I'm still exploring, many fun options! :)

**Comment by raoulcollenteur** [comment]
> @raoulcollenteur, both are working for me in jupyter lab. Which versions are you on? For me. the following worked:
> 
> * plotly 5.16.1
> * bokeh 3.4.0
> * jupyter lab 4.1.1
> 
> FYI, there seems to be some funkiness going on in the bokeh plots: image
> 
> And a note for anyone trying to view the bokeh plots in a VS Code notebook environment, the interactive figures don't show up. [bokeh/bokeh#10765](https://github.com/bokeh/bokeh/issues/10765)

Hmm, interesting... Was this in VS code or JupyterLab? Do you have [bokeh-jupyter-bokeh](https://github.com/bokeh/jupyter_bokeh) extension installed in JupyterLab? Perhaps that is necessary?

**Comment by dbrakenhoff** [comment]
> Was this in VS code or JupyterLab? Do you have [bokeh-jupyter-bokeh](https://github.com/bokeh/jupyter_bokeh) extension installed in JupyterLab? Perhaps that is necessary?

This was in JupyterLab. Both without and with the bokeh-jupyter-bokeh jupyterlab extension installed. 
VS code does not render Bokeh plots unfortunately. 

I think the data isn't sorted properly when I run `ml.get_output_series()`:

image

**Comment by raoulcollenteur** [comment]
Hmm, so the problem is not in bokeh plots. If I run the code above my data is well ordered.

Python version: 3.11.3
NumPy version: 1.24.4
Pandas version: 2.1.4
SciPy version: 1.12.0
Matplotlib version: None
Numba version: 0.59.0
LMfit version: 1.2.2
Latexify version: Not Installed
Pastas version: 1.5.0b

Is this a new Pandas problem?!

**Comment by dbrakenhoff** [comment]
That could well be it, pandas 2.2 here.

**Comment by raoulcollenteur** [comment]
I think it is related to this issue: https://github.com/pandas-dev/pandas/issues/57006

Probably good to fix in ml.get_out_series anyway.. Could you check if adding sort=True in pd.concat fixes the problem?

**Comment by dbrakenhoff** [comment]
Yep, that fixes it! Committed the change. Is it still working on your end?

**Comment by raoulcollenteur** [comment]
Yes all good here! I'll go through the changes in 2.2 again to see if more might be up. Should we merge this PR now?

**Review by dbrakenhoff — APPROVED** [review]

**Pull request merged**